docs/content/sips/xxx-manifest-redesign.md

```diff
@@ -0,0 +1,123 @@
+title = "SIP xxx - Application Manifest Redesign"
+template = "main"
+date = "2022-05-20T13:22:30Z"
+---
+
+Summary: A new design for Application Manifests (`spin.toml`)
+
+Owner: lann.martin@fermyon.com
+
+Created: May 20, 2022
+
+## Background
+
+Spin's Application Manifest file (usually named `spin.toml`) describes a Spin application's metadata,
+components, and trigger configurations. The current design supports a single trigger type and combines
+component and trigger configuration with an implicit one-to-one relationship:
+
+```toml
+name = "hello-world"
+trigger = { type = "http", base = "/hello" }
+
+[[component]]
+id = "world"
+# ...
+[component.trigger]
+route = "/world"
+```
+
+In order to support future features, we would like to enable more flexible trigger configurations:
+
+- Multiple trigger types in an application
+- Multiple triggers associated with a single component
+- One trigger associated with multiple components
+
+## Proposal
+
+### Global trigger config
+
+In order to allow for multiple trigger types, we remove the top-level `trigger.type` field.
+We also move all application metadata from the top-level config to a new `[application]`
+section, and all "global trigger config" to new `[application.trigger.<type>]` sections:
+
+```toml
+[application]
+name = "hello-world"
+[application.trigger.http]
+base = "/hello"
+```
+
+> In the short term we can continue to enforce the use of exactly one trigger type in an
+> application as a manifest validation step.
+
+### Trigger config
+
+In order to decouple triggers and components, we move triggers to new top-level
+`[[trigger.<type>]]` sections:
+
+```toml
+[[trigger.http]]
```

Comment:
I'm not sure I see the value of this over putting the trigger-related config under the component.  It adds a layer of indirection which means:

1. Two places to edit the component ID
2. When I look at the component it's not obvious what it's hooked up to (e.g. HTTP route)

Reply:
A couple of motivations:

- I think it makes more sense conceptually for triggers to be a top-level entity. Another design I considered was just inverting the component/trigger relationship, i.e. turning `component.trigger` into `trigger.component`, but:
- I want to enable HTTP middleware (and other kinds of trigger-component relationships), where a single trigger may reference multiple components (one "handler" and zero+ "interceptors"). I like the consistency of referencing them all by ID.

It's true that it introduces some duplication and potential non-locality, but in practice you can group a trigger and its "main" component together in the file, e.g.:

```toml
[[trigger.http]]
route = "/hello"
component = "world"
[[component]]
id = "world"
```

Reply:
ref #519

Reply:
@itowlson's perspective resonates with me because I have been thinking about Spin as a framework for building components and so components are the central subject in my head and everything revolves around them. However, when I tell myself that the spin.toml's job is to wire up triggers to components, then making the triggers a top level entity makes much more sense to me.